Webui use toolbar for local disks

rvykydal opened the pull request.

rvykydal commented:
Original screen:
![toolbox0](https://user-images.githubusercontent.com/1220237/185428571-b9314dba-85a7-405e-bc54-33d2c8af6112.png)
The new screen:
![toolbox1](https://user-images.githubusercontent.com/1220237/185428626-4626638a-a393-40cc-9cd8-53950b4f6dab.png)
![toolbox2](https://user-images.githubusercontent.com/1220237/185428661-3bec3660-40ee-4ae3-9bd7-0eb25454f877.png)
![toolbox3](https://user-images.githubusercontent.com/1220237/185428674-e84acb11-d72d-4f69-a357-2ea8ae66ab58.png)
![toolbox4](https://user-images.githubusercontent.com/1220237/185428683-b2ba364d-fad5-4137-ae5f-3a90219b2d15.png)
![toolbox5](https://user-images.githubusercontent.com/1220237/185428692-b539b285-91b7-4878-b9c7-0680c1e2aeea.png)

rvykydal commented:
I am aware of two issues:
- The help icons should be better aligned with "Local standard disks" label
- The separator between "1 selected" and the "Discover disks" button should be better placed.

poncovka commented on ui/webui/src/components/storage/InstallationDestination.jsx:
```diff
@@ -165,35 +173,26 @@ const LocalStandardDisks = ({ idPrefix, setIsFormValid, onAddErrorNotification }
+              })
+                      .finally(() => { setIsDiscoveringDisks(false) });
+              setIsDiscoveringDisks(true);
+              setSelectionForAllDisks(false);
```
Shoudn't we set the state before we start the discovery?

rvykydal replied:
Done.

poncovka commented on ui/webui/src/components/storage/InstallationDestination.jsx:
```diff
@@ -233,27 +232,99 @@ const LocalStandardDisks = ({ idPrefix, setIsFormValid, onAddErrorNotification }
+        setDisks(Object.keys(disks).reduce((acc, cur) => ({ ...acc, [cur]: value }), {}));
+    };
+
+    const DropdownSplitButtonText = () => {
```
It would be good to move this component on the top level of the file.

rvykydal replied:
Hm, I didn't think it really makes sense as a generic component outside the context of the local disks... But maybe when I actually implement the bulk logic it would make.

rvykydal replied:
Done.

poncovka commented on ui/webui/src/components/storage/InstallationDestination.jsx:
```diff
@@ -233,27 +232,99 @@ const LocalStandardDisks = ({ idPrefix, setIsFormValid, onAddErrorNotification }
+              toggle={
+                  <DropdownToggle
+                    splitButtonItems={[
+                        <DropdownToggleCheckbox
```
Please, define the checkbox as a separate variable if possible (`const dropdownCheckbox = ... `).

rvykydal replied:
Done.

poncovka commented on ui/webui/src/components/storage/InstallationDestination.jsx:
```diff
@@ -233,27 +232,99 @@ const LocalStandardDisks = ({ idPrefix, setIsFormValid, onAddErrorNotification }
+                            <ToolbarItem>{rescanDisksButton}</ToolbarItem>
+                        </ToolbarContent>
+                    </Toolbar>
+                    <ListingTable
```
Maybe the toolbar and the table could be also defined separately? It is just that every reorganization of the elements requires to move all these blocks of code even when they are not changed at all.

rvykydal replied:
Fixed.

poncovka reviewed (commented):
I have some suggestions. Also, the checkbox probably needs an additional logic, see my comment in the mockup.

rvykydal commented on ui/webui/src/components/storage/InstallationDestination.jsx:
```diff
@@ -233,27 +232,99 @@ const LocalStandardDisks = ({ idPrefix, setIsFormValid, onAddErrorNotification }
+              title={
+                  <Flex spaceItems={{ default: "spaceItemsXs" }}>
+                      <FlexItem><h3>{_("Local standard disks")}</h3></FlexItem>
                       <FlexItem>{localDisksInfo}</FlexItem>
```
I'll check it.

rvykydal replied:
I was able to get the best (though perhaps not ideal) result with PF Button in the PR update.
I am inclined to try to work on further improvements of this as a separate issue.

rvykydal commented:
> I have some suggestions. Also, the checkbox probably needs an additional logic, see my comment in the mockup.

Yes, I guess I'll need to implement https://www.patternfly.org/v4/guidelines/bulk-selection/#bulk-selection

VladimirSlavik commented on ui/webui/src/components/storage/InstallationDestination.jsx:
```diff
@@ -233,27 +232,99 @@ const LocalStandardDisks = ({ idPrefix, setIsFormValid, onAddErrorNotification }
+                          aria-label={_("Select multiple disks")}
+                          isChecked={selectedDisksCnt > 0}
+                        >
+                            {selectedDisksCnt > 0 ? cockpit.format(_("$0 selected"), selectedDisksCnt) : ""}
```
This needs a plural translation. I think the following would work:
```
{selectedDisksCnt > 0 ? cockpit.format(cockpit.ngettext("$0 selected", "$0 selected", selectedDisksCnt), selectedDisksCnt) : ""}
```

In other words, replace `_("$0 selected")` with `cockpit.ngettext("$0 selected", "$0 selected", selectedDisksCnt)`

See also: https://cockpit-project.org/guide/latest/cockpit-locale.html

rvykydal replied:
Yes, thank you, silly me was somehow stuck in english only here.

rvykydal replied:
Fixed

VladimirSlavik reviewed (commented):
I see only a few string cosmetics in addition to the other reviews.

rvykydal commented on ui/webui/src/components/storage/InstallationDestination.jsx:
```diff
@@ -233,27 +231,99 @@ const LocalStandardDisks = ({ idPrefix, setIsFormValid, onAddErrorNotification }
+            <DropdownItem
+              key="select-none"
+              component="button"
+              aria-label={_("Select none disk")}
```
Updated

rvykydal commented:
New update addressing reviews by Vendy and Vlada:
![Screen Shot 2022-08-25 at 11 26 38](https://user-images.githubusercontent.com/1220237/186633241-4b2090b2-6298-4183-b0f1-397752a01564.png)
![Screen Shot 2022-08-25 at 11 26 47](https://user-images.githubusercontent.com/1220237/186633248-4594e861-7584-40fd-b314-73fb01801c76.png)
![Screen Shot 2022-08-25 at 11 27 00](https://user-images.githubusercontent.com/1220237/186633262-d447319e-7490-45b4-8a58-c4f1f78df3ab.png)
![Screen Shot 2022-08-25 at 11 27 08](https://user-images.githubusercontent.com/1220237/186633273-635d9fcd-9520-4146-929e-37bff360defe.png)
![Screen Shot 2022-08-25 at 11 31 32](https://user-images.githubusercontent.com/1220237/186633284-a36c3fb6-c036-4ebc-b27c-dba45a5174c3.png)
![Screen Shot 2022-08-25 at 11 27 16](https://user-images.githubusercontent.com/1220237/186633300-dc0d63e5-46bb-478c-896b-21c6ad94b326.png)
![Screen Shot 2022-08-25 at 11 27 34](https://user-images.githubusercontent.com/1220237/186633312-d8029c90-3ab2-4129-bcea-f8cf5d8b68b9.png)

VladimirSlavik commented on ui/webui/src/components/storage/InstallationDestination.jsx:
```diff
@@ -233,27 +318,58 @@ const LocalStandardDisks = ({ idPrefix, setIsFormValid, onAddErrorNotification }
         }
     ));
 
+    const dropdownBulkSelect = (
+        <DropdownBulkSelect
```
I find it somewhat confusing to have two things with same name differing only in first letter upper/lower case.

VladimirSlavik reviewed (approved):
I think this is fine. Thank you!

M4rtinK commented:
Re-triggering cockpit CI now that it should be operational again. :)

M4rtinK commented:
/kickstart-test --testtype smoke

rvykydal commented on ui/webui/src/components/storage/InstallationDestination.jsx:
```diff
@@ -157,43 +244,41 @@ const LocalStandardDisks = ({ idPrefix, setIsFormValid, onAddErrorNotification }
+          variant="secondary"
           onClick={() => {
+              setIsDiscoveringDisks(true);
+              setDisks(setSelectionForAllDisks({ disks, value: false }));
```
Oh yes you are right. It needs more work.

M4rtinK reviewed (approved):
Looks good to me as well, thanks! :)

rvykydal commented:
Updated the pixel test reference.
